Structs with shared fields can't be used in string interpolation/println

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
OS: macos, macOS, 11.4, 20F71
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 12.0.5 (clang-1205.0.22.11)
vexe mtime: 2021-07-14 19:28:29
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 9c710b2.646c1e1

Git version: git version 2.30.1 (Apple Git-130)
Git vroot status: weekly.2021.24-267-g646c1e15 (9 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-unknown-unknown de82a130


**What did you do?**
Trying to print out the value of an instance of the App struct that has a shared field

```v
struct App {
	id string
mut:
	app_data shared AppData
}

struct AppData {
	id string
	sub_app_data SubAppData
}

struct SubAppData {
	id string
	message string = 'Default message'
}

fn main() {
	mut app := App {}
	println(app)
}
```

**What did you expect to see?**
Compiled app that generates the below (generated when `shared` keyword is removed from app_data field):
```js
App{
    id: ''
    app_data: AppData{
        id: ''
        sub_app_data: SubAppData{
            id: ''
            message: 'Default message'
        }
    }
}
```
 
**What did you see instead?**
Compile time error:
```bash
==================
                                                                                                           ~~ ^
/tmp/v/shared_test.13805284055148754337.tmp.c:2517:139: error: no member named 'sub_app_data' in 'struct __shared__main__AppData'
                {_SLIT("\n"), 0xfe10, {.d_s=indents}}, {_SLIT("    sub_app_data: "), 0, {.d_c=0}}, {_SLIT(""), 16, {.d_s=indent_main__SubAppData_str(it.sub_app_data, indent_count + 1)}}, {_SLIT(""), 0, {.d_c=0}},
                                                                                                                                                     ~~ ^
2 errors generated.
...
==================
```

With latest V, there's only one errors, and it's different from avove...
```
/tmp/v/shared.15505899162754791322.tmp.c:2583: error: field not found: id
builder error: 
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5360,8 +5360,7 @@ fn (mut g Gen) type_default(typ_ ast.Type) string {
 			mut has_none_zero := false
 			mut init_str := '{'
 			info := sym.info as ast.Struct
-			typ_is_shared_f := typ.has_flag(.shared_f)
-			if sym.language == .v && !typ_is_shared_f {
+			if sym.language == .v {
 				for field in info.fields {
 					field_sym := g.table.sym(field.typ)
 					if field.has_default_expr
@@ -5392,21 +5391,24 @@ fn (mut g Gen) type_default(typ_ ast.Type) string {
 					}
 				}
 			}
+			typ_is_shared_f := typ.has_flag(.shared_f)
 			if has_none_zero {
 				init_str += '}'
-				type_name := if info.is_anon {
-					// No name needed for anon structs, C figures it out on its own.
-					''
-				} else {
-					'(${g.typ(typ)})'
+				if !typ_is_shared_f {
+					type_name := if info.is_anon {
+						// No name needed for anon structs, C figures it out on its own.
+						''
+					} else {
+						'(${g.typ(typ)})'
+					}
+					init_str = type_name + init_str
 				}
-				init_str = type_name + init_str
 			} else {
 				init_str += '0}'
 			}
-			if typ.has_flag(.shared_f) {
+			if typ_is_shared_f {
 				styp := '__shared__${g.table.sym(typ).cname}'
-				return '($styp*)__dup${styp}(&($styp){.mtx = {0}, .val =$init_str}, sizeof($styp))'
+				return '($styp*)__dup${styp}(&($styp){.mtx = {0}, .val = $init_str}, sizeof($styp))'
 			} else {
 				return init_str
 			}

--- a/vlib/v/tests/struct_shared_field_with_default_init_test.v
+++ b/vlib/v/tests/struct_shared_field_with_default_init_test.v
@@ -1,0 +1,23 @@
+struct App {
+	id string
+mut:
+	app_data shared AppData
+}
+
+struct AppData {
+	id           string
+	sub_app_data SubAppData
+}
+
+struct SubAppData {
+	id      string
+	message string = 'Default message'
+}
+
+fn test_struct_field_with_default_init() {
+	mut app := App{}
+	println(app)
+	rlock app.app_data {
+		assert app.app_data.sub_app_data.message == 'Default message'
+	}
+}

